File: JavaScript/752-Open the Lock.js
/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
var openLock = function(deadends, target) {
    let deadendset = new Set(deadends);
    // edge case
    if(deadendset.has('0000')) return -1;
    deadendset.add('0000');
    
    let queue = [['0000',0]];
    
    function* findNxt(v) {
        for (let i=0; i<v.length; i++){
            const num = Number(v[i]);
            for (let change of [1, -1]){
                yield v.slice(0,i) + ((num+change)%10+10)%10  + v.slice(i+1);  
            }
        }
    }
    
    while (queue.length){
        const size = queue.length;
        for (let i=0; i<size; i++){
            let [curr, count] = queue.shift();
            if (curr === target) return count;
            let generator = findNxt(curr);
            while (true){
                let nxt = generator.next().value;
                if (!nxt) break;
                if (deadendset.has(nxt)) continue;
                deadendset.add(nxt);
                queue.push([nxt, count+1]);
            }
        }
    }
    return -1
};

File: C++/752-Open the Lock.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        // BFS
        unordered_set<string> visited(deadends.begin(), deadends.end());
        if (visited.find("0000") != visited.end()) return -1;
        
        visited.insert("0000");
        queue<string> q;
        q.push("0000");
        int step = 0;
        
        while (!q.empty()) {
            int length = q.size();
            while (length > 0) {
                string curr = q.front();
                q.pop();
                if (curr == target) return step;
                for (int i=0; i<4; i++) {
                    char tmp = curr[i];
                    curr[i] = (curr[i] - '0' + 1)%10 + '0';
                    if (visited.find(curr) == visited.end()) {
                        visited.insert(curr);
                        q.push(curr);
                    }
                    curr[i] = tmp;
                }
                for (int i=0; i<4; i++) {
                    char tmp = curr[i];
                    curr[i] = (curr[i] - '0' - 1 + 10)%10 + '0';
                    if (visited.find(curr) == visited.end()) {
                        visited.insert(curr);
                        q.push(curr);
                    }
                    curr[i] = tmp;
                }
                length--;
            }
            step++;
        }
        return -1;
    }
};